### ISSUE TITLE
Add ability to programatically define a file route

### ISSUE THREAD

**Issue by OrangeTimes** [vikejs/vike#1691]
### Description

Current Vike API doesn't provide any way how to programmatically create a file route for situations when +Page.ts are created externally or in other dynamic fashhion

**Comment by brillout** [comment]
Use case: https://github.com/vikejs/vike/discussions/1661.

@OrangeTimes I don't think there is any React/Vue/Solid(/...) framework out there that supports this?

**Comment by OrangeTimes** [comment]
Yes, Next.Js and other similar frameworks don't support this at the moment

**Comment by doeixd** [comment]
This is possible with Solid Start and Nuxt.

**Comment by brillout** [comment]
If you know some links to their docs about it that would be great.

**Comment by doeixd** [comment]
I edited the post with some docs links. Sorry

**Comment by brillout** [comment]
Thank you!

**Comment by OrangeTimes** [comment]
> This is possible with Solid Start and Nuxt. I think
> 
> https://nuxt.com/docs/guide/recipes/custom-routing
> 
> https://docs.solidjs.com/solid-start/building-your-application/routing#additional-route-config (FilesystemRoutes component returns an object which you can add to)
> 
> https://remix.run/docs/en/main/discussion/routes#manual-route-configuration

Correct me if I'm wrong,  but in Remix it will be evaluated one time and not on each request. If I need to push a new file route on some incoming request condition (create a file route on the fly) this will not work most probably.

**Comment by doeixd** [comment]
Thats correct. 

For your use case, I'd suggest just using a client side router (Whose routes can be added to on the server / client) + a backend service that you write that will ssr the pages manually.

**Comment by OrangeTimes** [comment]
We currently have a very custom and quite old SSR server. If there was a way to create file routes on the fly using onBeforeRoute for example it would drastically simplify things for us and our existing setup and would unblock us from adopting Vike potentially.

**Comment by srghma** [comment]
👍 would allow to define routes using https://github.com/natefaubion/purescript-routing-duplex

**Comment by brillout** [comment]
An idea would be to have routes be defined at `pageContext.routes` as a single source of truth that users can manipulate. Use case example:
- https://github.com/vikejs/vike/discussions/3076#discussioncomment-15675755

Although one issue is that `pageContext.routes` needs to the same on both server- and client side... so not sure how feasible this idea is.

**Comment by doeixd** [comment]
Astro just announced a feature similar to this. Just gonna leave this here incase its helpful. 

[https://astro.build/blog/astro-6-beta/#live-collections-stable](https://astro.build/blog/astro-6-beta/#live-collections-stable)

**Comment by brillout** [comment]
@doeixd This ticket is more about programmatically defining routes, while AFAIK Astro's live collections are more about data fetching (with type safety and nice DX). Feel free to create new discussions and/or feature requests, if you want something Astro has that Vike doesn't. There are many things to like about Astro's DX, and it's on the radar to cherry-pick some of their ideas.

**Comment by doeixd** [comment]
Sorry if it wasn't useful. I genuinely thought the functionality was very similar, because you can use the live collections to generate routes, after the application has already been deployed. 

From the docs:

<img width="773" height="143" alt="Image" src="https://github.com/user-attachments/assets/bd87f411-2ed5-44d5-8842-77d5ce71c706" />

**Comment by brillout** [comment]
👍 I didn't know about this, although it seems their requirements are quite different (e.g. because Astro doesn't do Client-Side Routing).

Ideally, a mutable `pageContext.routes` would be great. But I ain't sure how we can ensure it's synced between the client- and server-side.

There is distinction to be made here between programmatically defined at *config time* (e.g. inside `+config.js`) or programmatically defined at *runtime* (i.e. routes can be changed post build-time).

Programmatic at config time is easy and definitely on the radar (I'll actually personally need it for a couple of things I want to implement).

Programmatic at runtime is tricky, so I'm inclined to postpone this a bit. That said, one neat use case for it is using CMS data to generate pages (CMS data can dynamically change at runtime).

**Comment by brillout** [comment]
Done:
 - https://github.com/vikejs/vike/pull/3356

Pre-released as `0.4.259-commit-f6ecfb2`.

As for doing it at runtime, I created a new feature request:
- https://github.com/vikejs/vike/issues/3359
